LiveIntervals: Use BumpPtrAllocator

arsenm opened the pull request.

arsenm committed:
LiveIntervals: Use BumpPtrAllocator

arsenm commented:
* **#127057** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/127057?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/127057?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

jayfoad commented on llvm/include/llvm/CodeGen/LiveIntervals.h:
```diff
@@ -64,6 +65,8 @@ class LiveIntervals {
   MachineDominatorTree *DomTree = nullptr;
   std::unique_ptr<LiveIntervalCalc> LICalc;
 
+  BumpPtrAllocator Allocator;
```
Comment saying what this is used for?

jayfoad reviewed (commented):
Seems reasonable but do you have any performance data? Have you tried it with llvm-compile-time-tracker.com? And does it increase memory usage measurably, since BumpPtrAllocator::Deallocate does nothing?

arsenm commented:
> Seems reasonable but do you have any performance data? 

No, I did this months ago and wanted to find something, but forgot about it and figured the tracker is the easiest way to test. I don't actually know how to submit branches there. @nikic?

nikic reviewed (commented):
LiveRange is not POD, so isn't this going to leak big time?

You need to use SpecificBumpPtrAllocator if you want to run destructors.

arsenm committed:
Use SpecificBumpPtrAllocator for LiveInterval

I didn't realize we used a singly linked list for storing subranges,
but that seems bad.

I didn't realize we used a singly linked list for storing subranges.
That seems bad and we should probably switch this to an array

nikic commented:
With the new commit memory usage is a better but still +16% on clang: http://llvm-compile-time-tracker.com/compare.php?from=8600d89e55b866186a2dad2e2b4c85bd96150375&to=bd8b3bc6f30b78b51fde4b7fae7bcd67963117fb&stat=max-rss

nikic commented on llvm/include/llvm/CodeGen/LiveIntervals.h:
```diff
@@ -64,6 +65,12 @@ class LiveIntervals {
+  BumpPtrAllocator Allocator;
+
+  // Allocator for VirtRegIntervals
+  SpecificBumpPtrAllocator<LiveInterval> LIAllocator;
```
Need this for LiveRange as well, as it contains vectors.

arsenm committed:
Use yet another allocator for LiveRanges

Not sure it's worth it for these, there should never be all that
many. We could pre-allocate the maximum size up front.
